test/Numerics/LinearSolvers/bgmres.jl

```diff
@@ -0,0 +1,244 @@
+x = ArrayType(randn(n, ni)) # initial guess
+A = ArrayType(randn((n, n, ni)) ./ sqrt(n) .* 1.0)
+for i in 1:n
+    A[i,i,:] .+= 1.0
+end
+ss = size(b)[1]
+gmres = BatchedGeneralizedMinimalResidual(b, ArrayType = ArrayType, subspace_size = ss)
+for i in 1:ni
+    x[:,i] = A[:, :, i] \ b[:, i]
+end
+sol = copy(x)
+x += ArrayType(randn(n,ni) * 0.01 * maximum(abs.(x)))
+y = copy(x)
+linear_operator! = closure_linear_operator_multi!(A, size(A)...)
+iters = linearsolve!(linear_operator!, gmres, x, b; max_iters = ss)
+linear_operator!(y, x)
+@testset "(CPU) Basic Test" begin
+    @test norm(y-b) / norm(b) < err_thresh
+end
+
+if CUDAapi.has_cuda_gpu()
+    ArrayType = CuArray
+    n  = 100  # size of vector space
+    ni = 10 # number of independent linear solves
+    b = ArrayType(randn(n, ni)) # rhs
+    x = ArrayType(randn(n, ni)) # initial guess
+    A = ArrayType(randn((n, n, ni)) ./ sqrt(n) .* 1.0)
+    for i in 1:n
+        A[i,i,:] .+= 1.0
+    end
+    ss = size(b)[1]
+    gmres = BatchedGeneralizedMinimalResidual(b, ArrayType = ArrayType, subspace_size = ss)
+    for i in 1:ni
+        x[:,i] = A[:, :, i] \ b[:, i]
+    end
+    sol = copy(x)
+    x += ArrayType(randn(n,ni) * 0.01 * maximum(abs.(x)))
+    y = copy(x)
+    linear_operator! = closure_linear_operator_multi!(A, size(A)...)
+    iters = linearsolve!(linear_operator!, gmres, x, b; max_iters = ss)
+    linear_operator!(y, x)
+    @testset "(GPU) Basic Test" begin
+        @test norm(y-b) / norm(b) < err_thresh
+    end
+end
+###
+# Test 2: MPIStateArray test (CPU)
+ArrayType = Array
+Random.seed!(1235)
+n1 = 8
+n2 = 3
+n3 = 10
+mpi_b = MPIStateArray{T}(mpicomm, ArrayType, n1, n2, n3)
+mpi_x = MPIStateArray{T}(mpicomm, ArrayType, n1, n2, n3)
+mpi_A = ArrayType(randn(n1*n2, n1*n2, n3))
+
+# need to make sure that mpi_b and mpi_x are reproducible
+mpi_b.data[:] .= ArrayType(randn(n1 * n2 * n3))
+mpi_x.data[:] .= ArrayType(randn(n1 * n2 * n3))
+mpi_y = copy(mpi_x)
+
+# for defining linear_operator
+function closure_linear_operator_mpi!(A, n1, n2, n3)
+    function linear_operator!(x, y)
+        alias_x = reshape(x.data,(n1, n3))
+        alias_y = reshape(y.data,(n1, n3))
+        event = multiply_by_A!(alias_x, A, alias_y, n1, n2, ndrange = n3)
+        wait(event)
+        return nothing
+    end
+end
+
+gmres = BatchedGeneralizedMinimalResidual(mpi_b, ArrayType = ArrayType, m = n1*n2, n = n3)
+
+# Now define the linear operator
+linear_operator! = closure_linear_operator_mpi!(mpi_A, size(mpi_A)...)
+iters = linearsolve!(linear_operator!, gmres, mpi_x, mpi_b; max_iters = n1*n2)
+linear_operator!(mpi_y, mpi_x)
+# check one in the batch
+sol = mpi_A[:,:,1] \ mpi_b.data[:,:, 1][:]
+@testset "(CPU) MPIStateArray Test" begin
+    @test norm(mpi_y.data - mpi_b.data) / norm(mpi_b) < err_thresh
+    @test norm(sol - mpi_x.data[:,:,1][:]) < err_thresh
+end
+
+if CUDAapi.has_cuda_gpu()
+    ArrayType = CuArray
+    Random.seed!(1235)
+    n1 = 8
+    n2 = 3
+    n3 = 10
```

Comment:
For all three of these tests, it looks like some variables (like these) are defined twice. This could be cleaned up a bit by creating an array of `[Array, CuArray]` if a gpu is present or just `[Array]` if not.

Reply:
good idea